File: custom_msg_srv/src/test_custom_interfaces/test_custom_interfaces/subscriber_member_function.py
#! /usr/bin/env python3

import rclpy
from rclpy.node import Node
from custom_msg_srv.msg import Num
from rclpy.executors import ExternalShutdownException


class MinimalSubscriber(Node):
    def __init__(self):
        super().__init__("num_subscriber")

        self._subscription = self.create_subscription(Num,
                                                      "my_num",
                                                      self._listener_callback,
                                                      10
                                                      )

    def _listener_callback(self, msg):
        self.get_logger().info(f'I heard: "{msg.num}"')


def main(args=None):
    try:
        rclpy.init(args=args)
        node = MinimalSubscriber()
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        print()
    except Exception as e:
        print(f"An error occurred: {e}")
    finally:
        if node is not None:
            node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()

File: custom_msg_srv/src/test_custom_interfaces/src/subscriber_member_function.cpp
#include <functional>
#include <memory>
#include "rclcpp/rclcpp.hpp" // ROS 2 C++ Client Lib
#include "custom_msg_srv/msg/num.hpp" // Handling String messages

using std::placeholders::_1; //Placeholder for callback function

class MinimalSubscriber : public rclcpp::Node {
  public:
    MinimalSubscriber() : Node("num_subscriber"){
      subscription_ = this->create_subscription<custom_msg_srv::msg::Num>(
        "my_num", 10, std::bind(&MinimalSubscriber::topic_callback, this, _1));
    }

  private:
    void topic_callback(const custom_msg_srv::msg::Num & msg) const {
      RCLCPP_INFO(this->get_logger(), "I heard: %ld", msg.num);
    }
    rclcpp::Subscription<custom_msg_srv::msg::Num>::SharedPtr subscription_;
};

int main(int argc, char* argv[]){
  rclcpp::init(argc, argv);
  auto node = std::make_shared<MinimalSubscriber>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
